File: examples/anthropic-managed-agents/javascript/src/webhook-runtime.ts
import Anthropic from "@anthropic-ai/sdk";
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import {
  createWriteStream,
  existsSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { spawn } from "node:child_process";

const REMOTE_DIR = "/opt/anthropic-managed-agents-js";
const REMOTE_WORKDIR = "/mnt/session";
const REMOTE_TSX = `${REMOTE_DIR}/node_modules/.bin/tsx`;
const REMOTE_WORKER = `${REMOTE_DIR}/src/worker-runtime.ts`;
const REMOTE_PID = `${REMOTE_DIR}/worker.pid`;
const REMOTE_PIDS_DIR = `${REMOTE_DIR}/worker-pids`;
const REMOTE_LOG = `${REMOTE_DIR}/worker.log`;
const REMOTE_ENVIRONMENT_ID = `${REMOTE_WORKDIR}/.anthropic-environment-id`;
const REMOTE_ENVIRONMENT_KEY = `${REMOTE_WORKDIR}/.anthropic-environment-key`;
const REMOTE_WEBHOOK_SIGNING_KEY = `${REMOTE_WORKDIR}/.anthropic-webhook-signing-key`;
const REMOTE_WORKER_MAX_IDLE_SECONDS = `${REMOTE_WORKDIR}/.worker-max-idle-seconds`;
const REMOTE_LOG_LEVEL = `${REMOTE_WORKDIR}/.log-level`;
const MAX_WEBHOOK_BODY_BYTES = 1_048_576;
const MAX_WORKERS = Number(process.env.MAX_WORKERS ?? "4");
const WORKER_RETRY_MS = 5_000;

const client = new Anthropic({ apiKey: "not-needed" });
let pendingWorkerStarts = 0;
let workerRetryTimer: NodeJS.Timeout | undefined;

class PayloadTooLargeError extends Error {}

function fileValue(path: string) {
  if (!existsSync(path)) {
    return undefined;
  }
  const value = readFileSync(path, "utf8").trim();
  return value || undefined;
}

function configValue(name: string, path: string) {
  return process.env[name] || fileValue(path);
}

function workerEnv() {
  return {
    ANTHROPIC_ENVIRONMENT_ID: configValue("ANTHROPIC_ENVIRONMENT_ID", REMOTE_ENVIRONMENT_ID),
    ANTHROPIC_ENVIRONMENT_KEY: configValue("ANTHROPIC_ENVIRONMENT_KEY", REMOTE_ENVIRONMENT_KEY),
    WORKER_MAX_IDLE_SECONDS: configValue("WORKER_MAX_IDLE_SECONDS", REMOTE_WORKER_MAX_IDLE_SECONDS),
    WORKER_RUN_SECONDS: process.env.WORKER_RUN_SECONDS,
    LOG_LEVEL: configValue("LOG_LEVEL", REMOTE_LOG_LEVEL),
    PATH: process.env.PATH,
    HOME: process.env.HOME,
  };
}

function processIsRunning(pid: number) {
  return Number.isInteger(pid) && existsSync(`/proc/${pid}`);
}

function activeWorkerPids() {
  mkdirSync(REMOTE_PIDS_DIR, { recursive: true });
  const pids: number[] = [];

  for (const file of readdirSync(REMOTE_PIDS_DIR)) {
    const path = `${REMOTE_PIDS_DIR}/${file}`;
    const pid = Number(readFileSync(path, "utf8").trim());
    if (processIsRunning(pid)) {
      pids.push(pid);
    } else {
      unlinkSync(path);
    }
  }

  if (existsSync(REMOTE_PID)) {
    const pid = Number(readFileSync(REMOTE_PID, "utf8").trim());
    if (processIsRunning(pid) && !pids.includes(pid)) {
      pids.push(pid);
    }
  }

  return pids;
}

function scheduleWorkerRetry() {
  if (workerRetryTimer || pendingWorkerStarts === 0) {
    return;
  }

  workerRetryTimer = setTimeout(() => {
    workerRetryTimer = undefined;
    startWorkerIfCapacity({ retryingPendingStart: true });
  }, WORKER_RETRY_MS);
}

function startWorkerIfCapacity({ retryingPendingStart = false } = {}) {
  const pids = activeWorkerPids();
  if (pids.length >= MAX_WORKERS) {
    if (!retryingPendingStart) {
      pendingWorkerStarts = Math.min(pendingWorkerStarts + 1, MAX_WORKERS);
    }
    scheduleWorkerRetry();
    return;
  }

  if (retryingPendingStart) {
    pendingWorkerStarts = Math.max(0, pendingWorkerStarts - 1);
  }

  const log = createWriteStream(REMOTE_LOG, { flags: "a" });
  const child = spawn(REMOTE_TSX, [REMOTE_WORKER], {
    cwd: REMOTE_WORKDIR,
    detached: true,
    env: workerEnv(),
    stdio: ["ignore", "pipe", "pipe"],
  });
  child.stdout.pipe(log);
  child.stderr.pipe(log);
  child.unref();
  writeFileSync(REMOTE_PID, `${child.pid}\n`);
  writeFileSync(`${REMOTE_PIDS_DIR}/${child.pid}.pid`, `${child.pid}\n`);
  child.on("exit", () => {
    try {
      unlinkSync(`${REMOTE_PIDS_DIR}/${child.pid}.pid`);
    } catch {
      // The health check can clean stale pid files too.
    }
    scheduleWorkerRetry();
  });
}

function webhookSigningKey() {
  const keyPath = process.env.ANTHROPIC_WEBHOOK_SIGNING_KEY_FILE ?? REMOTE_WEBHOOK_SIGNING_KEY;
  return configValue("ANTHROPIC_WEBHOOK_SIGNING_KEY", keyPath);
}

async function readBody(request: IncomingMessage, maxBytes: number) {
  const contentLength = request.headers["content-length"];
  if (contentLength && Number(contentLength) > maxBytes) {
    throw new PayloadTooLargeError("request body too large");
  }

  const chunks: Buffer[] = [];
  let size = 0;
  for await (const chunk of request) {
    const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
    size += buffer.length;
    if (size > maxBytes) {
      throw new PayloadTooLargeError("request body too large");
    }
    chunks.push(buffer);
  }
  return Buffer.concat(chunks).toString("utf8");
}

function writeJson(response: ServerResponse, statusCode: number, body: unknown) {
  response.writeHead(statusCode, { "content-type": "application/json" });
  response.end(JSON.stringify(body));
}

async function handleWebhook(request: IncomingMessage, response: ServerResponse) {
  const signingKey = webhookSigningKey();
  if (!signingKey) {
    response.writeHead(503, { "content-type": "text/plain" });
    response.end("ANTHROPIC_WEBHOOK_SIGNING_KEY is required");
    return;
  }

  let body: string;
  try {
    body = await readBody(request, MAX_WEBHOOK_BODY_BYTES);
  } catch (error) {
    if (error instanceof PayloadTooLargeError) {
      response.writeHead(413, { "content-type": "text/plain" });
      response.end("request body too large");
      return;
    }
    throw error;
  }

  try {
    const event = client.beta.webhooks.unwrap(body, {
      headers: Object.fromEntries(
        Object.entries(request.headers).map(([key, value]) => [
          key,
          Array.isArray(value) ? value.join(", ") : (value ?? ""),
        ]),
      ),
      key: signingKey,
    });

    if (event.data.type === "session.status_run_started") {
      startWorkerIfCapacity();
    }

    response.writeHead(204);
    response.end();
  } catch {
    response.writeHead(401, { "content-type": "text/plain" });
    response.end("invalid signature");
  }
}

const server = createServer((request, response) => {
  if (request.method === "GET" && request.url === "/health") {
    const pids = activeWorkerPids();
    writeJson(response, 200, { ok: true, worker_running: pids.length > 0, worker_count: pids.length });
    return;
  }

  if (request.method === "POST" && request.url === "/webhook") {
    void handleWebhook(request, response).catch((error) => {
      console.error(error);
      if (!response.headersSent) {
        response.writeHead(500, { "content-type": "text/plain" });
      }
      response.end("webhook handler failed");
    });
    return;
  }

  response.writeHead(404, { "content-type": "text/plain" });
  response.end("not found");
});

const port = Number(process.env.WEBHOOK_PORT ?? "8000");
server.listen(port, "0.0.0.0", () => {
  console.log(`webhook server listening on ${port}`);
});

File: examples/anthropic-managed-agents/python/anthropic_managed_agents_e2b/webhook_runtime.py
from __future__ import annotations

import os
import subprocess
import threading
from pathlib import Path

import anthropic
from fastapi import FastAPI, Request, Response

REMOTE_DIR = Path("/opt/anthropic-managed-agents")
REMOTE_WORKDIR = Path("/mnt/session")
REMOTE_WORKER = REMOTE_DIR / "worker.py"
REMOTE_PID = REMOTE_DIR / "worker.pid"
REMOTE_PIDS_DIR = REMOTE_DIR / "worker-pids"
REMOTE_LOG = REMOTE_DIR / "worker.log"
REMOTE_ENVIRONMENT_ID = REMOTE_WORKDIR / ".anthropic-environment-id"
REMOTE_ENVIRONMENT_KEY = REMOTE_WORKDIR / ".anthropic-environment-key"
REMOTE_WEBHOOK_SIGNING_KEY = REMOTE_WORKDIR / ".anthropic-webhook-signing-key"
REMOTE_WORKER_MAX_IDLE_SECONDS = REMOTE_WORKDIR / ".worker-max-idle-seconds"
REMOTE_LOG_LEVEL = REMOTE_WORKDIR / ".log-level"
MAX_WEBHOOK_BODY_BYTES = 1_048_576
MAX_WORKERS = max(1, int(os.environ.get("MAX_WORKERS", "4")))
WORKER_RETRY_SECONDS = 5

app = FastAPI()
client = anthropic.Anthropic(api_key="not-needed")
pending_worker_starts = 0
worker_retry_timer: threading.Timer | None = None


def file_value(path: Path) -> str | None:
    if not path.exists():
        return None
    value = path.read_text().strip()
    return value or None


def config_value(name: str, path: Path) -> str | None:
    return os.environ.get(name) or file_value(path)


def worker_env() -> dict[str, str]:
    env = {
        "ANTHROPIC_ENVIRONMENT_ID": config_value(
            "ANTHROPIC_ENVIRONMENT_ID",
            REMOTE_ENVIRONMENT_ID,
        ),
        "ANTHROPIC_ENVIRONMENT_KEY": config_value(
            "ANTHROPIC_ENVIRONMENT_KEY",
            REMOTE_ENVIRONMENT_KEY,
        ),
        "WORKER_MAX_IDLE_SECONDS": config_value(
            "WORKER_MAX_IDLE_SECONDS",
            REMOTE_WORKER_MAX_IDLE_SECONDS,
        ),
        "WORKER_RUN_SECONDS": os.environ.get("WORKER_RUN_SECONDS"),
        "LOG_LEVEL": config_value("LOG_LEVEL", REMOTE_LOG_LEVEL),
        "PATH": os.environ.get("PATH"),
        "HOME": os.environ.get("HOME"),
    }
    return {key: value for key, value in env.items() if value is not None}


def process_is_running(pid: int) -> bool:
    return pid > 0 and Path(f"/proc/{pid}").exists()


def pid_file_value(path: Path) -> int | None:
    try:
        return int(path.read_text().strip())
    except (OSError, ValueError):
        return None


def active_worker_pids() -> list[int]:
    REMOTE_PIDS_DIR.mkdir(parents=True, exist_ok=True)
    pids: list[int] = []

    for path in REMOTE_PIDS_DIR.iterdir():
        pid = pid_file_value(path)
        if pid is not None and process_is_running(pid):
            pids.append(pid)
        else:
            path.unlink(missing_ok=True)

    latest_pid = pid_file_value(REMOTE_PID)
    if latest_pid is not None and process_is_running(latest_pid) and latest_pid not in pids:
        pids.append(latest_pid)

    return pids


def schedule_worker_retry() -> None:
    global worker_retry_timer

    if worker_retry_timer is not None or pending_worker_starts == 0:
        return

    worker_retry_timer = threading.Timer(
        WORKER_RETRY_SECONDS,
        lambda: start_worker_if_capacity(retrying_pending_start=True),
    )
    worker_retry_timer.daemon = True
    worker_retry_timer.start()


def cleanup_worker_on_exit(process: subprocess.Popen[bytes]) -> None:
    process.wait()
    (REMOTE_PIDS_DIR / f"{process.pid}.pid").unlink(missing_ok=True)
    schedule_worker_retry()


def start_worker_if_capacity(*, retrying_pending_start: bool = False) -> None:
    global pending_worker_starts, worker_retry_timer

    if retrying_pending_start:
        worker_retry_timer = None

    if len(active_worker_pids()) >= MAX_WORKERS:
        if not retrying_pending_start:
            pending_worker_starts = min(pending_worker_starts + 1, MAX_WORKERS)
        schedule_worker_retry()
        return

    if retrying_pending_start:
        pending_worker_starts = max(0, pending_worker_starts - 1)

    with REMOTE_LOG.open("ab") as log_file:
        process = subprocess.Popen(
            ["python", str(REMOTE_WORKER)],
            cwd=REMOTE_WORKDIR,
            stdin=subprocess.DEVNULL,
            stdout=log_file,
            stderr=subprocess.STDOUT,
            start_new_session=True,
            env=worker_env(),
        )
    REMOTE_PID.write_text(f"{process.pid}\n")
    (REMOTE_PIDS_DIR / f"{process.pid}.pid").write_text(f"{process.pid}\n")
    threading.Thread(target=cleanup_worker_on_exit, args=(process,), daemon=True).start()
    schedule_worker_retry()


def webhook_signing_key() -> str | None:
    key_file = Path(
        os.environ.get("ANTHROPIC_WEBHOOK_SIGNING_KEY_FILE", REMOTE_WEBHOOK_SIGNING_KEY)
    )
    return config_value("ANTHROPIC_WEBHOOK_SIGNING_KEY", key_file)


async def read_limited_body(request: Request, max_bytes: int) -> str:
    content_length = request.headers.get("content-length")
    if content_length and int(content_length) > max_bytes:
        raise ValueError("request body too large")

    chunks: list[bytes] = []
    size = 0
    async for chunk in request.stream():
        size += len(chunk)
        if size > max_bytes:
            raise ValueError("request body too large")
        chunks.append(chunk)
    return b"".join(chunks).decode()


@app.get("/health")
def health() -> dict[str, bool | int]:
    pids = active_worker_pids()
    return {"ok": True, "worker_running": len(pids) > 0, "worker_count": len(pids)}


@app.post("/webhook")
async def webhook(request: Request) -> Response:
    signing_key = webhook_signing_key()
    if not signing_key:
        return Response("ANTHROPIC_WEBHOOK_SIGNING_KEY is required", status_code=503)

    try:
        payload = await read_limited_body(request, MAX_WEBHOOK_BODY_BYTES)
    except ValueError:
        return Response("request body too large", status_code=413)

    try:
        event = client.beta.webhooks.unwrap(
            payload,
            headers=dict(request.headers),
            key=signing_key,
        )
    except Exception:
        return Response("invalid signature", status_code=401)

    if event.data.type == "session.status_run_started":
        start_worker_if_capacity()

    return Response(status_code=204)
